lib/os/onoff.c

```diff
@@ -0,0 +1,522 @@
+	}
+
+	k_spin_unlock(&srv->lock, key);
+
+	/* Notify the releaser.  If there was an error, notify any
+	 * pending requests; otherwise if there are pending requests
+	 * start the transition to ON.
+	 */
+	notify_one(srv, releaser, res);
+	if (notify_clients) {
+		notify_all(srv, &clients, client_res);
+	} else if (start) {
+		srv->start(srv, onoff_start_notify);
+	}
+}
+
+int onoff_release(struct onoff_service *srv,
+		  struct onoff_client *cli)
+{
+	bool stop = false;      /* invoke stop transition */
+	bool notify = false;    /* do client notification */
+	int rv = validate_args(srv, cli);
+
+	if (rv < 0) {
+		return rv;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) != 0) {
+		rv = -EIO;
+		goto out;
+	}
+
+	u32_t state = srv->flags & SERVICE_STATE_MASK;
+
+	switch (state) {
+	case SERVICE_STATE_ON:
+		/* Stay on if release leaves a client. */
+		if (srv->refs > 1U) {
+			notify = true;
+			rv = 1;
+			break;
+		}
+
+		/* Reject if in non-thread context but stop could
+		 * wait
+		 */
+		if ((k_is_in_isr() || k_is_pre_kernel())
+		    && ((srv->flags & ONOFF_SERVICE_STOP_SLEEPS) != 0)) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		stop = true;
+
+		set_service_state(srv, SERVICE_STATE_TO_OFF);
+		srv->releaser = cli;
+		rv = 2;
+
+		break;
+	case SERVICE_STATE_TO_ON:
+		rv = -EBUSY;
+		break;
+	case SERVICE_STATE_OFF:
+	case SERVICE_STATE_TO_OFF:
+		rv = -EALREADY;
+		break;
+	default:
+		rv = -EINVAL;
+	}
+
+out:
+	if (notify) {
+		srv->refs -= 1U;
+	}
+
+	k_spin_unlock(&srv->lock, key);
+
+	if (stop) {
+		__ASSERT_NO_MSG(srv->stop != NULL);
+		srv->stop(srv, onoff_stop_notify);
+	} else if (notify) {
+		notify_one(srv, cli, 0);
+	}
+
+	return rv;
+}
+
+static void onoff_reset_notify(struct onoff_service *srv,
+			       int res)
+{
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+	sys_slist_t clients = srv->clients;
+
+	/* If the reset failed clear the transition flag but otherwise
+	 * leave the state unchanged.
+	 *
+	 * If it was successful clear the reference count and all
+	 * flags except capability flags (sets to SERVICE_STATE_OFF).
+	 */
+	if (res < 0) {
+		srv->flags &= ~SERVICE_STATE_TRANSITION;
+	} else {
+		__ASSERT_NO_MSG(srv->refs == 0U);
+		srv->refs = 0U;
+		srv->flags &= SERVICE_CONFIG_FLAGS;
+	}
+
+	sys_slist_init(&srv->clients);
+
+	k_spin_unlock(&srv->lock, key);
+
+	notify_all(srv, &clients, res);
+}
+
+int onoff_service_reset(struct onoff_service *srv,
+			struct onoff_client *cli)
+{
+	if (srv->reset == NULL) {
+		return -ENOTSUP;
+	}
+
+	bool reset = false;
+	int rv = validate_args(srv, cli);
+
+	if (rv < 0) {
+		return rv;
+	}
+
+	/* Reject if in a non-thread context and reset could wait. */
+	if ((k_is_in_isr() || k_is_pre_kernel())
+	    && ((srv->flags & ONOFF_SERVICE_RESET_SLEEPS) != 0U)) {
+		return -EWOULDBLOCK;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) == 0) {
+		rv = -EINVAL;
```

Comment:
what about having `-EALREADY` here? Since if user do according to the book and tries reset on failure then it indicates that someone else already performed reset.